### PR TITLE
Adds Dockerfile, docker-compose and dev instructions on readme

### PR DESCRIPTION
**Pull request by kamiradi**
This PR adds 

- A development Dockerfile
- Corresponding docker-compose
- Necessary workflow instruction on the readme

### CODE REVIEW

File: .docker/Dockerfile
Comment (sea-bass) [inline .docker/Dockerfile:41]: nit: But it makes things a little more readable and consistent to always use the braces, so `${WORKSPACE}` instead of `$WORKSPACE`

File: .docker/docker-compose.yaml
Comment (sea-bass) [inline .docker/docker-compose.yaml:12]: Suggest also adding `QT_X11_NO_MITSHM=1` to the env vars to get display working

**Review by sea-bass — CHANGES_REQUESTED** [review]
Looks good! Almost worked on my end, but I was missing some `rosdep` dependencies of `moveit_drake`:

```
  Could not find a package configuration file provided by
  "moveit_visual_tools" with any of the following names:

    moveit_visual_toolsConfig.cmake
    moveit_visual_tools-config.cmake
```

Might be useful to actually do a `COPY` of the source code during build and run `rosdep install` with that `package.xml` also included at build.

Although @sjahr I notice that installing the `moveit_visual_tools` dependency ends up installing a bunch of moveit binaries, which is not ideal. Can we maybe remove some of the dependencies in that `package.xml` separately? Maybe this pipeline benchmark demo is adding a lot of dependencies we don't want?

Even after that, inside the Docker I get some build errors:

```
CMake Error at demo/CMakeLists.txt:1 (add_executable):
  Target "pipeline_testbench_example" links to target
  "moveit_core::moveit_butterworth_parameters" but the target was not found.
  Perhaps a find_package() call is missing for an IMPORTED target, or an
  ALIAS target is missing?


CMake Error at demo/CMakeLists.txt:1 (add_executable):
  Target "pipeline_testbench_example" links to target
  "moveit_core::moveit_planning_request_adapter" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?
```

---

Besides the build failures from the existing code that was in this repo, I think there are some easy improvements to make -- see the inline comments for details. But overall:

* First, I would try to consolidate some of the installs, as there seem to be a few places where the same commands are running multiple times, which shouldn't be necessary.
* I also think if there is a way to not install MoveIt packages from apt in the first place so you don't have to uninstall them, that would be preferable.
* Last blocker is that there seem to be some environment variables that don't seem to exist, which likely are an effect of borrowing this setup from somewhere else.

---

As a comment for later, it seems like this Docker setup won't persist the output of `colcon build` on the workspace, which is something I am happy to help modify myself after this PR is merged. Definitely not needed for this PR.

File: README.md
Comment (sea-bass) [inline README.md:23]: pro tip, this one is agnostic to the container changing names per instance

```
    docker compose exec -it moveit bash
```

**Comment by sea-bass** [comment]
Actually, several of those issues building were due to some other misconfigurations I found. You guys should try my PR https://github.com/kamiradi/moveit_drake/pull/1 and merge it into this branch if it works.

**Comment by kamiradi** [comment]
Thanks for the review @sea-bass. A few overall comments on the docker build and how I intend to use the image/container, let me know if I am missing something.

> Looks good! Almost worked on my end, but I was missing some `rosdep` dependencies of `moveit_drake`:
> 
> ```
>   Could not find a package configuration file provided by
>   "moveit_visual_tools" with any of the following names:
> 
>     moveit_visual_toolsConfig.cmake
>     moveit_visual_tools-config.cmake
> ```
> Might be useful to actually do a `COPY` of the source code during build and run `rosdep install` with that `package.xml` also included at build.

The Docker image just ensures a drake installation and built moveit packages. After that, you will need to start the container and perform the following steps that were part of the initial build instructions suggested by @sjahr 

1. Perform the `moveit-drake` vcs import.
2. Do the necessary rosdep installation
3. `colcon build`

I put these in the [readme](https://github.com/kamiradi/moveit_drake/tree/dev?tab=readme-ov-file#docker-workflow-preferred-and-tested). Didn't face any issues with the build. I did not copy the repository using `COPY . .` because I usually mount the repository using the compose file, purely for dev reasons. Obviously this can be optimized by just ensuring `moveit_drake` also builds with the rest of `moveit2`. Just thought that since I may be anyway building `moveit_drake` as part of the dev process, I'll keep it separate.

> Although @sjahr I notice that installing the `moveit_visual_tools` dependency ends up installing a bunch of moveit binaries, which is not ideal. Can we maybe remove some of the dependencies in that `package.xml` separately? Maybe this pipeline benchmark demo is adding a lot of dependencies we don't want?
> 
> Even after that, inside the Docker I get some build errors:
> 
> ```
> CMake Error at demo/CMakeLists.txt:1 (add_executable):
>   Target "pipeline_testbench_example" links to target
>   "moveit_core::moveit_butterworth_parameters" but the target was not found.
>   Perhaps a find_package() call is missing for an IMPORTED target, or an
>   ALIAS target is missing?
> 
> 
> CMake Error at demo/CMakeLists.txt:1 (add_executable):
>   Target "pipeline_testbench_example" links to target
>   "moveit_core::moveit_planning_request_adapter" but the target was not
>   found.  Perhaps a find_package() call is missing for an IMPORTED target, or
>   an ALIAS target is missing?
> ```
Hmmm, I didn't come across this. Does this stop the build or pop up as a warning?

> Besides the build failures from the existing code in this repo, I think there are some easy improvements to make -- see the inline comments for details. But overall:
> 
> * First, I would try to consolidate some of the installs, as there seem to be a few places where the same commands are running multiple times, which shouldn't be necessary.
> * I also think if there is a way to not install MoveIt packages from apt in the first place so you don't have to uninstall them, that would be preferable.
> * Last blocker is that there seem to be some environment variables that don't seem to exist, which likely are an effect of borrowing this setup from somewhere else.
Strong agree on these. I'll take all your docker optimizations Thank you! :smiley: 

> 
> As a comment for later, it seems like this Docker setup won't persist the output of `colcon build` on the workspace, which is something I am happy to help modify myself after this PR is merged. Definitely not needed for this PR.
This again is more for dev reasons than anything else.

**Comment by kamiradi** [comment]
> Actually, several of those issues building the code were due to some other misconfigurations I found. You guys should try my PR [kamiradi#1](https://github.com/kamiradi/moveit_drake/pull/1) and merge it into this branch if it works.
> 
> It fixes everything except these errors in that `demo` subfolder:
> 
> ```
> CMake Error at demo/CMakeLists.txt:1 (add_executable):
>   Target "pipeline_testbench_example" links to target
>   "moveit_core::moveit_butterworth_parameters" but the target was not found.
>   Perhaps a find_package() call is missing for an IMPORTED target, or an
>   ALIAS target is missing?
> 
> 
> CMake Error at demo/CMakeLists.txt:1 (add_executable):
>   Target "pipeline_testbench_example" links to target
>   "moveit_core::moveit_planning_request_adapter" but the target was not
>   found.  Perhaps a find_package() call is missing for an IMPORTED target, or
>   an ALIAS target is missing?
> ```

LGTM, I'll wait for @sjahr to review before commiting merging it

**Review by sjahr — APPROVED** [review]
Docker builds for me, nice work!

> Although @sjahr I notice that installing the moveit_visual_tools dependency ends up installing a bunch of moveit binaries, which is not ideal. Can we maybe remove some of the dependencies in that package.xml separately? Maybe this pipeline benchmark demo is adding a lot of dependencies we don't want?

Good question, for me running `sudo apt purge ros-humble-moveit*` before building the workspace fixed the build error above. I know it is not the prettiest solution but maybe that is the fastest way to fix this for now. We'll need moveit_visual_tools for the demo unfortunately :see_no_evil: @kamiradi Do you mind adding that to this PR and then we'll merge it

**Comment by sea-bass** [comment]
> Good question, for me running `sudo apt purge ros-humble-moveit*` before building the workspace fixed the build error above. I know it is not the prettiest solution but maybe that is the fastest way to fix this for now. We'll need moveit_visual_tools for the demo unfortunately 🙈 @kamiradi Do you mind adding that to this PR and then we'll merge it

Gah sorry, @kamiradi had that in originally and I removed it thinking it was harmless.

**Comment by sjahr** [comment]
:sweat_smile: Oh damn

File: .docker/docker-compose.yaml
Comment (kamiradi) [reply .docker/docker-compose.yaml]: I forgot to mention the steps of setting the variable. I wasn't able to create/find the .Xauthority file in my home directory. Is this something that has to be explicitly created?

**Comment by kamiradi** [comment]
I have made the changes, but haven't built the image yet as I am facing some Xauthority issues when trying to the new way. @sea-bass, can you point me in the direction of how I can create the `$HOME/.Xauthority` file?

**Comment by sjahr** [comment]
:thinking: Adding the command to the docker file does not seem to fix the issue. When I re-build the container and then run colcon build inside it, is still see the issue. Running the remove command again, fixes it. Also for some reason, not all repositories from the .repos file are cloned into the src folder. My folder in the docker just contains
`generate_parameter_library  geometric_shapes  moveit2  moveit_drake  moveit_msgs  moveit_resources`

**Comment by sjahr** [comment]
@kamiradi I am intrigued to merge this PR, and open issues to fix this. Let's talk about the problems in our meeting later today and move forward. None of the issues are blocking I think

**Comment by kamiradi** [comment]
In the current 

> 🤔 Adding the command to the docker file does not seem to fix the issue. When I re-build the container and then run colcon build inside it, is still see the issue. Running the remove command again, fixes it. Also for some reason, not all repositories from the .repos file are cloned into the src folder. My folder in the docker just contains `generate_parameter_library geometric_shapes moveit2 moveit_drake moveit_msgs moveit_resources`

Yes indeed, in my original version of the PR, I intended the `moveit_drake` build instructions on the readme to stay. After building the image, you will still need to do an additional `moveit_drake` `vcs import` before updating and installing the `rosdeps`, and then build the repository.

**Pull request merged**
